File: arrays_strings/sliding-window/max_sum_subarray.ts
function maxSumSubArray(arr: number[], k: number): number {
  const n = arr.length;

  // check for the edge case
  if (n < k) {
    return -1;
  }

  let windowSum = 0;

  // calculate initial window sum for k
  for (let i = 0; i < k; i++) {
    windowSum += arr[i];
  }

  let maxSum = windowSum;

  // slide the window to the right
  for (let i = k; i < n; i++) {
    windowSum += arr[i] - arr[i - k];
    if (maxSum < windowSum) {
      maxSum = windowSum;
    }
  }

  return maxSum;
}

const arr_num: number[] = [2, 1, 5, 1, 3, 2];
const k: number = 3;
console.log(maxSumSubArray(arr_num, k));

File: arrays_strings/sliding-window/max_sum_subarray.py
def maxSumSubArray(nums, k):
    n = len(nums)
    window_sum = 0
    max_sum = 0

    if n < k:
        return -1

    # calculate the sum of first window size K
    for i in range(k):
        window_sum += nums[i]

    max_sum = window_sum

    # slide the window from start to the end of the array

    for i in range(k, n):
        window_sum += nums[i] - nums[i - k]

        if window_sum > max_sum:
            max_sum = window_sum

    return max_sum


nums = [2, 1, 5, 1, 3, 2]
k = 3

print(maxSumSubArray(nums, k))
